lib/serialization/examples/custom_serialization.nit

```diff
@@ -0,0 +1,155 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Example of an ad hoc serializer that is tailored to transform business specific objects into customized representation.
+#
+# In the following, we expose 3 business classes, `E`, `A` and `B`, and a specific business serializer `RestrictedSerializer`.
+# The principle is that the custom serialization logic in enclosed into the `RestrictedSerializer` and that the
+# standard serializer is unchanged.
+#
+# The additional behaviors exposed are:
+#
+# * replace a full business object. (see `E::id`).
+#   instead of serializing an attribute, the custom serializer use a different representation
+# * inject a phantom attribute (see `E::css_class`).
+#   when serializing, the custom serializer inject new attributes.
+# * hide a normally serialized attribute (see `E::semi_private`)
+#   when serializing, the custom serializer hides some specific attributes.
+#
+# The advantage of the approach is that it is done programmatically so can be adapted to real complex use cases.
+# Basically, this is half-way between the full automatic serialization and the full manual serialisation.
+module custom_serialization
+
+import serialization
+import json::serialization_write
+
+# The root class of the business objects.
+# This factorizes most services and domain knowledge used by the `RestrictedSerializer`
+#
+# In real enterprise-level code, the various specific behaviors can be specified in more semantic classifiers.
+abstract class E
+	serialize
+
+	# The semantic business identifier.
+	#
+	# With the `RestrictedSerializer`, references to `E` objects will be replaced with `id`-based information.
+	# This avoid to duplicate or enlarge the information cross-call wise.
+	#
+	# A future API/REST call can then request the _missing_ object from its identifier.
+	var id: String
+
+	# A phantom attribute to be serialized by the custom `RestrictedSerializer`.
+	#
+	# This can be used to inject constant or computed information that make little sense to have as a genuine attribute in
+	# the Nit model.
+	fun css_class: String do return "So Much Fun"
```

Comment:
If you add spaces in the class, maybe the attribute name should be `css_classes`